File: bluetooth/wifi/wifi-state.js
var events = require('events');
var util = require('util');
var mqtt = require('mqtt');

function WifiState() {
    events.EventEmitter.call(this);

    this.ssid = "";
    this.psk = "";
    this.last_attempt = false;

    this.clientId = 'wifi_bt_peripheral';
    this.has_received_first_update;

    var that = this;
    var client_connection_topic = 'wifi/connection/' + this.clientId + '/state';
    this.wifi_publish_topic = 'wifi/set_config';
    this.wifi_subscribe_topic = 'wifi/response';

    var mqtt_options = {
        clientID: this.clientID,
        'will' : {
            topic: client_connection_topic,
            payload: '0',
            qos: 2,
            retain: true,
        },
    }

    var mqtt_client = mqtt.connect('mqtt://localhost', mqtt_options);

    mqtt_client.on('connect', function() {
        console.log("Connected to mqtt broker!")
        mqtt_client.publish(client_connection_topic,
            '1', {qos:2, retain:true})
        mqtt_client.subscribe(that.wifi_subscribe_topic);
    });

    mqtt_client.on('message', function(topic, message) {
        if (topic === that.wifi_subscribe_topic) {
            let new_data = JSON.parse(message);
            if (new_data['status'] == true) {
                that.last_attempt = true;
            } else {
                that.last_attempt = false;
            }

            console.log(new_data);

            that.emit('received-update-response', that.last_attempt);
        }
        console.log("MQTT Message: ", topic, message);
    });

    this.mqtt_client = mqtt_client;
}

util.inherits(WifiState, events.EventEmitter);

WifiState.prototype.set_ssid = function(ssid) {
    console.log("Setting state ssid to: ", ssid);
    this.ssid = ssid;
}

WifiState.prototype.set_psk = function(psk) {
    console.log("Setting state psk to: ", psk);
    this.psk = psk;
}

WifiState.prototype.join_wifi = function() {
    if (this.ssid != "") {
        var tmp = {'client' : this.clientId, 'ssid' : this.ssid}
        if (this.psk != "") {
            tmp['psk'] = this.psk;
        }
        this.mqtt_client.publish(this.wifi_publish_topic, JSON.stringify(tmp), {qos:1});
        console.log("Set wifi data");
        this.last_attempt = true;
        return true;
    } else {
        console.log("Missing SSID");
        this.mqtt_client.publish(this.wifi_subscribe_topic, JSON.stringify({'client': "",
                  'status': false}), {qos:1});
        this.last_attempt = false;
        return false;
    }
}

module.exports = WifiState;

File: bluetooth/wifi/wifi_state.py
import sys
import json

TOPIC_WIFI_PUBLISH = 'wifi/set_config'
TOPIC_WIFI_SUBSCRIBE = 'wifi/response'


class WifiState:
    def __init__(self, helper):
        self._helper = helper
        self._mqtt_client = helper.mqtt_client
        self._client = self._mqtt_client.client

        self.ssid = ""
        self.psk = ""
        self.last_attempt = False
        self.last_attempt_feedback = "TODO: IMPLEMENT FEEDBACK"
        self.has_received_first_update = False

        self.update_action_callback = None
        self.feedback_callback = None

        self._setup_mqtt_callbacks()

    def _setup_mqtt_callbacks(self):
        self._client.message_callback_add(TOPIC_WIFI_SUBSCRIBE,
                                          self.on_receive_wifi_message)
        self._client.subscribe(TOPIC_WIFI_SUBSCRIBE, qos=1)

    def on_receive_wifi_message(self, client, userdata, msg):
        new_data = json.loads(msg.payload.decode('utf-8'))
        print(new_data)

        if 'status' not in new_data:
            #TODO: throw error
            return

        if new_data['status'] is True:
            self.last_attempt = True
        else:
            self.last_attempt = False

        if self.update_action_callback is not None:
            self.update_action_callback(self.last_attempt)

        if self.feedback_callback is not None:
            self.feedback_callback(self.last_attempt_feedback)

    def set_ssid(self, ssid):
        self.ssid = ssid

    def set_psk(self, psk):
        self.psk = psk

    def join_wifi(self):
        if not self.ssid:
            print("Missing SSID")
            self.last_attempt = False
            return

        msg = {'ssid': self.ssid}
        if self.psk:
            msg['psk'] = self.psk
        self._mqtt_client.publish_dict(TOPIC_WIFI_PUBLISH, msg, qos=1)
        print("set wifi data")





